modules/pam_shells/pam_shells.c

```diff
@@ -48,6 +50,85 @@ static int perform_check(pam_handle_t *pamh)
     if (userShell[0] == '\0')
 	userShell = DEFAULT_SHELL;
 
+#if defined (USE_ECONF)	&& defined (VENDORDIR)
+    size_t size = 0;
+    econf_file **key_files, **begin_key_files;
+    econf_err error;
+    char **keys;
+    econf_file *key_file;
+
+    error = econf_readDirsHistory(&key_files,
+				  &size,
+				  VENDORDIR,
+				  ETCDIR,
+				  SHELLS,
+				  NULL, /* no suffix */
+				  "", /* key only */
+				  "#"  /* comment */);
+    if (error) {
+	pam_syslog(pamh, LOG_ERR,
+		   "Cannot parse shell files: %s",
+		   econf_errString(error));
+	return PAM_AUTH_ERR;
+    }
+
+    begin_key_files = key_files;
+    while(*key_files) {
+	char *file_path =  econf_getPath (*key_files);
+	if (stat(file_path, &sb)) {
+	    pam_syslog(pamh, LOG_ERR, "Cannot stat %s: %m", file_path);
+	    free(file_path);
+	    free(begin_key_files);
+	    return PAM_AUTH_ERR;
+	}
+
+	if ((sb.st_mode & S_IWOTH) || !S_ISREG(sb.st_mode)) {
+	    pam_syslog(pamh, LOG_ERR,
+	        "%s is either world writable or not a normal file",
+		file_path);
+	    free(file_path);
+	    free(begin_key_files);
+	    return PAM_AUTH_ERR;
+	}
+	free(file_path);
+        key_files++;
+    }
+    free(begin_key_files);
+
+    error = econf_readDirs(&key_file,
```

Comment:
Here `econf_readDirs` is called, which in turn would invoke `econf_readDirsHistory` again, but this time without all the checks implemented above, which is obviously a TOCTOU race condition.
Regretfully, econf doesn't provide a race-free API for this task yet.
What do you think about extending econf API, e,g, with a variant of `econf_readDirs` with a callback function?

Reply:
That's an good idea! I have implemented it.